File: watcher/argo.py
import json
import logging
from time import time
from typing import Optional

import requests
from prometheus_client import Gauge
from requests.exceptions import RequestException
from requests.packages.urllib3.exceptions import InsecureRequestWarning
from tenacity import RetryError
from tenacity import retry
from tenacity import retry_if_exception_type
from tenacity import stop_after_delay
from tenacity import wait_fixed

from watcher.models import Task
from watcher.settings import Settings
from watcher.state import DBState
from watcher.state import InMemoryState

match Settings.Watcher.state_type:
    case "in-memory":
        state = InMemoryState()
    case "postgres":
        state = DBState()
    case _:
        logging.error("Invalid STATE_TYPE was provided")
        exit(1)


class InvalidImageException(Exception):
    pass


class AppNotReadyException(Exception):
    pass


class AppDoesNotExistException(Exception):
    pass


class Argo:
    def __init__(self):
        self.session = requests.Session()
        self.session.verify = Settings.Watcher.ssl_verify
        if not self.session.verify:
            # Reducing noise in logs as we assume that the user knows
            # the risk of disabling SSL verification
            requests.packages.urllib3.disable_warnings(InsecureRequestWarning)

        # Exposes per-app Prometheus metric in the following way
        # failed_deployment{app_name="test"} 0.0
        self.failed_deployment_gauge = Gauge(
            "failed_deployment", "Failed deployment", ["app_name"]
        )

        self.argo_url = Settings.Argo.url
        self.argo_user = Settings.Argo.user
        self.argo_password = Settings.Argo.password
        self.authorized = self.auth()

    def auth(self) -> bool:
        url = f"{self.argo_url}/api/v1/session"
        payload = {"username": self.argo_user, "password": self.argo_password}

        try:
            response = self.session.post(url=url, json=payload)
        except RequestException as exception:
            logging.error(exception)
            return False

        match response.status_code:
            case 401:
                logging.critical("Unauthorized, please check ArgoCD credentials!")
                exit(1)
            case 403:
                logging.critical("Forbidden, please check the firewall!")
                exit(1)

    def check_argo(self):
        try:
            response = self.session.get(url=f"{self.argo_url}/api/v1/session/userinfo")
            if response.json()["loggedIn"]:
                return "up"
        except KeyError:
            return "down"

    def start_task(self, task: Task):
        logging.info(
            f"New task with id {task.id} was triggered. "
            f"Expecting tag {task.images[0].tag} in {task.app} app."
        )
        try:
            state.set_current_task(task=task, status="in progress")
            self.wait_for_rollout(task=task)
            self.failed_deployment_gauge.labels(task.app).set(0)
            state.update_task(task_id=task.id, status="deployed")
            logging.info(
                f"Task {task.id} has succeeded. "
                f"App {task.app} is running on the expected version."
            )
        except RetryError:
            logging.warning(
                f"Task {task.id} has failed. "
                f"App {task.app} did not become healthy within reasonable timeframe."
            )
            state.update_task(task_id=task.id, status="failed")
            self.failed_deployment_gauge.labels(task.app).inc()
        except AppDoesNotExistException:
            logging.warning(
                f"Task {task.id} has failed. " f"App {task.app} does not exists."
            )
            state.update_task(task_id=task.id, status="app not found")

    @staticmethod
    def get_task_status(task_id: str):
        return state.get_task_status(task_id=task_id)

    @staticmethod
    def return_state(from_timestamp: float, to_timestamp: float, app_name: str):
        # Set "to_timestamp" to the current timestamp
        # to return all tasks starting from "from_timestamp"
        if to_timestamp is None:
            to_timestamp = time()

        return state.get_state(
            time_range_from=(time() - from_timestamp) / 60,
            time_range_to=to_timestamp / 60,
            app_name=app_name,
        )

    @staticmethod
    def return_app_list():
        return state.get_app_list()

    def refresh_app(self, app: str) -> int:
        # Trigger app refresh to increase new version detection speed
        url = f"{self.argo_url}/api/v1/applications/{app}?refresh=normal"
        return self.session.get(url=url).status_code

    def get_app_status(self, app: str) -> Optional[dict]:
        r = self.session.get(url=f"{self.argo_url}/api/v1/applications/{app}")
        if r.status_code != 200:
            return None
        else:
            app = json.loads(r.content.decode("utf-8"))

        status = {
            "images": app["status"]["summary"]["images"],
            "synced": app["status"]["sync"]["status"],
            "healthy": app["status"]["health"]["status"],
        }

        return status

    @retry(
        stop=stop_after_delay(Settings.Argo.timeout),
        retry=retry_if_exception_type((AppNotReadyException, InvalidImageException)),
        wait=wait_fixed(5),
    )
    def wait_for_rollout(self, task: Task):

        if self.refresh_app(app=task.app) == 404:
            raise AppDoesNotExistException

        app_status = self.get_app_status(task.app)

        for target in task.images:
            if f"{target.image}:{target.tag}" not in app_status["images"]:
                logging.debug(f"{target.image}:{target.tag} is not available yet...")
                raise InvalidImageException

        if app_status["synced"] == "Synced" and app_status["healthy"] == "Healthy":
            return
        else:
            raise AppNotReadyException

File: cmd/argo-watcher/argo.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/avast/retry-go/v4"
	"github.com/google/uuid"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/rs/zerolog/log"
	"io"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	s "github.com/shini4i/argo-watcher/cmd/argo-watcher/state"
	c "github.com/shini4i/argo-watcher/internal/config"
	h "github.com/shini4i/argo-watcher/internal/helpers"
	m "github.com/shini4i/argo-watcher/internal/models"
)

var (
	argoTimeout, _        = strconv.Atoi(h.GetEnv("ARGO_TIMEOUT", "0"))
	argoSyncRetryDelay    = 15 * time.Second
	retryAttempts         = uint((argoTimeout / 15) + 1)
	config                = c.GetConfig()
	argoPlannedRetryError = errors.New("planned retry")
	imagesProxy           = os.Getenv("DOCKER_IMAGES_PROXY")
)

const (
	ArgoAppSuccess = iota
	ArgoAppNotSynced
	ArgoAppNotAvailable
	ArgoAppNotHealthy
	ArgoAppFailed
)

const (
	ArgoAPIErrorTemplate        = "ArgoCD API Error: %s"
	argoUnavailableErrorMessage = "connect: connection refused"
)

type Argo struct {
	Url     string
	Token   string
	Timeout string
	client  *http.Client
	state   s.State
}

func (argo *Argo) Init() error {
	log.Debug().Msg("Initializing argo-watcher client...")

	switch state := os.Getenv("STATE_TYPE"); state {
	case "postgres":
		argo.state = &s.PostgresState{}
		argo.state.Connect()
	case "in-memory":
		argo.state = &s.InMemoryState{}
	default:
		log.Fatal().Msg("Variable STATE_TYPE must be one of [\"postgres\", \"in-memory\"]. Aborting.")
	}

	log.Debug().Msgf("Configured retry attempts per ArgoCD application status check: %d", retryAttempts)

	go argo.state.ProcessObsoleteTasks()

	skipTlsVerify, _ := strconv.ParseBool(h.GetEnv("SKIP_TLS_VERIFY", "false"))

	transport := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: skipTlsVerify},
	}

	jar, err := cookiejar.New(nil)
	if err != nil {
		return err
	}

	cookie := &http.Cookie{
		Name:  "argocd.token",
		Value: argo.Token,
	}

	argoUrl, err := url.Parse(argo.Url)
	if err != nil {
		return err
	}

	jar.SetCookies(argoUrl, []*http.Cookie{cookie})

	argoApiTimeout, _ := strconv.Atoi(argo.Timeout)
	argo.client = &http.Client{
		Transport: transport,
		Jar:       jar,
		Timeout:   time.Duration(argoApiTimeout) * time.Second,
	}

	log.Debug().Msgf("Timeout for ArgoCD API calls set to: %s", argo.client.Timeout)

	return nil
}

func (argo *Argo) Check() (string, error) {
	if argo == nil {
		argocdUnavailable.Set(1)
		return "", errors.New("argo-watcher is not initialized yet")
	}

	apiUrl := fmt.Sprintf("%s/api/v1/session/userinfo", argo.Url)
	req, err := http.NewRequest("GET", apiUrl, nil)
	if err != nil {
		panic(err)
	}

	req.Header.Set("Content-Type", "application/json; charset=UTF-8")

	resp, err := argo.client.Do(req)
	if err != nil {
		log.Error().Msg(err.Error())
		argocdUnavailable.Set(1)
		return config.StatusArgoCDUnavailableMessage, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}

	defer func(Body io.ReadCloser) {
		err := Body.Close()
		if err != nil {
			log.Error().Msg(err.Error())
		}
	}(resp.Body)

	var userInfo m.Userinfo
	if err = json.Unmarshal(body, &userInfo); err != nil {
		panic(err)
	}

	if argo.state.Check() && userInfo.LoggedIn {
		argocdUnavailable.Set(0)
		return "up", nil
	} else {
		// This error is misleading, need to introduce a new one
		argocdUnavailable.Set(1)
		return "down", errors.New(config.StatusArgoCDUnavailableMessage)
	}
}

func (argo *Argo) AddTask(task m.Task) (string, error) {
	status, err := argo.Check()
	if err != nil {
		return status, errors.New(err.Error())
	}

	task.Id = uuid.New().String()

	log.Info().Str("id", task.Id).Msgf("A new task was triggered. Expecting tag %s in app %s.",
		task.Images[0].Tag,
		task.App,
	)

	argo.state.Add(task)
	processedDeployments.Inc()
	go argo.waitForRollout(task)
	return task.Id, nil
}

func (argo *Argo) GetTasks(startTime float64, endTime float64, app string) m.TasksResponse {
	_, err := argo.Check()
	tasks := argo.state.GetTasks(startTime, endTime, app)

	if err != nil {
		return m.TasksResponse{
			Tasks: tasks,
			Error: err.Error(),
		}
	}

	return m.TasksResponse{
		Tasks: tasks,
	}
}

func (argo *Argo) checkAppStatus(app string) (*m.Application, error) {
	apiUrl := fmt.Sprintf("%s/api/v1/applications/%s", argo.Url, app)
	req, err := http.NewRequest("GET", apiUrl, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json; charset=UTF-8")

	q := req.URL.Query()
	q.Add("refresh", "normal")
	req.URL.RawQuery = q.Encode()

	resp, err := argo.client.Do(req)
	if err != nil {
		return nil, err
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	defer func(Body io.ReadCloser) {
		err := Body.Close()
		if err != nil {
			log.Error().Msg(err.Error())
		}
	}(resp.Body)

	if resp.StatusCode != 200 {
		var argoErrorResponse m.ArgoApiErrorResponse
		err = json.Unmarshal(body, &argoErrorResponse)
		if err != nil || argoErrorResponse.Message == "" {
			return nil, errors.New(bytes.NewBuffer(body).String())
		} else {
			return nil, errors.New(argoErrorResponse.Message)
		}
	}

	var argoApp m.Application
	err = json.Unmarshal(body, &argoApp)
	if err != nil {
		return nil, err
	}

	return &argoApp, nil
}

func (argo *Argo) checkWithRetry(task m.Task) (int, error) {
	var status int

	err := retry.Do(
		func() error {
			app, err := argo.checkAppStatus(task.App)

			if err != nil {
				status = ArgoAppFailed
				return err
			}

			for _, image := range task.Images {
				expected := fmt.Sprintf("%s:%s", image.Image, image.Tag)
				if !h.ImageContains(app.Status.Summary.Images, expected, imagesProxy) {
					log.Debug().Str("id", task.Id).Msgf("%s is not available yet", expected)
					status = ArgoAppNotAvailable
					return argoPlannedRetryError
				}
			}

			if app.Status.Sync.Status != "Synced" {
				log.Debug().Str("id", task.Id).Msgf("%s is not synced yet", task.App)
				status = ArgoAppNotSynced
				return argoPlannedRetryError
			}

			if app.Status.Health.Status != "Healthy" {
				log.Debug().Str("id", task.Id).Msgf("%s is not healthy yet", task.App)
				status = ArgoAppNotHealthy
				return argoPlannedRetryError
			}

			status = ArgoAppSuccess
			return nil
		},
		retry.DelayType(retry.FixedDelay),
		retry.Delay(argoSyncRetryDelay),
		retry.Attempts(retryAttempts),
		retry.RetryIf(func(err error) bool {
			return errors.Is(err, argoPlannedRetryError)
		}),
		retry.LastErrorOnly(true),
	)

	return status, err
}

func (argo *Argo) waitForRollout(task m.Task) {
	// continuously check for application status change
	status, err := argo.checkWithRetry(task)

	// application synced successfully
	if status == ArgoAppSuccess {
		argo.handleDeploymentSuccess(task)
		return
	}

	// we had some unexpected error with ArgoCD API
	if status == ArgoAppFailed {
		argo.handleArgoAPIFailure(task, err)
		return
	}

	// fetch application details
	app, err := argo.checkAppStatus(task.App)

	// define default message
	const defaultErrorMessage string = "could not retrieve details"
	// handle application sync failure
	switch status {
	// not all images were deployed to the application
	case ArgoAppNotAvailable:
		// show list of missing images
		var message string
		// define details
		if err != nil {
			message = defaultErrorMessage
		} else {
			message = "List of current images (last app check):\n"
			message += "\t" + strings.Join(app.Status.Summary.Images, "\n\t") + "\n\n"
			message += "List of expected images:\n"
			message += "\t" + strings.Join(task.ListImages(), "\n\t")
		}
		// handle error
		argo.handleAppNotAvailable(task, errors.New(message))
	// application sync status wasn't valid
	case ArgoAppNotSynced:
		// display sync status and last sync message
		var message string
		// define details
		if err != nil {
			message = defaultErrorMessage
		} else {
			message = "App status \"" + app.Status.OperationState.Phase + "\"\n"
			message += "App message \"" + app.Status.OperationState.Message + "\"\n"
			message += "Resources:\n"
			message += "\t" + strings.Join(app.ListSyncResultResources(), "\n\t")
		}
		// handle error
		argo.handleAppOutOfSync(task, errors.New(message))
	// application is not in a healthy status
	case ArgoAppNotHealthy:
		// display current health of pods
		var message string
		// define details
		if err != nil {
			message = defaultErrorMessage
		} else {
			message = "App sync status \"" + app.Status.Sync.Status + "\"\n"
			message += "App health status \"" + app.Status.Health.Status + "\"\n"
			message += "Resources:\n"
			message += "\t" + strings.Join(app.ListUnhealthyResources(), "\n\t")
		}
		// handle error
		argo.handleAppNotHealthy(task, errors.New(message))
	// handle unexpected status
	default:
		argo.handleDeploymentUnexpectedStatus(task, errors.New(fmt.Sprintf("Received unexpected status \"%d\"", status)))
	}
}

func (argo *Argo) handleArgoAPIFailure(task m.Task, err error) {
	// notify user that app wasn't found
	appNotFoundError := fmt.Sprintf("applications.argoproj.io \"%s\" not found", task.App)
	if strings.Contains(err.Error(), appNotFoundError) {
		argo.handleAppNotFound(task, err)
		return
	}
	// notify user that ArgoCD API isn't available
	if strings.Contains(err.Error(), argoUnavailableErrorMessage) {
		argo.handleArgoUnavailable(task, err)
		return
	}

	// notify of unexpected error
	argo.handleDeploymentFailed(task, err)
	return
}

func (argo *Argo) handleAppNotFound(task m.Task, err error) {
	log.Info().Str("id", task.Id).Msgf("Application %s does not exist.", task.App)
	reason := fmt.Sprintf(ArgoAPIErrorTemplate, err.Error())
	argo.state.SetTaskStatus(task.Id, config.StatusAppNotFoundMessage, reason)
}

func (argo *Argo) handleArgoUnavailable(task m.Task, err error) {
	log.Error().Str("id", task.Id).Msg("ArgoCD is not available. Aborting.")
	reason := fmt.Sprintf(ArgoAPIErrorTemplate, err.Error())
	argo.state.SetTaskStatus(task.Id, "aborted", reason)
}

func (argo *Argo) handleDeploymentFailed(task m.Task, err error) {
	log.Warn().Str("id", task.Id).Msgf("Deployment failed. Aborting with error: %s", err)
	failedDeployment.With(prometheus.Labels{"app": task.App}).Inc()
	reason := fmt.Sprintf(ArgoAPIErrorTemplate, err.Error())
	argo.state.SetTaskStatus(task.Id, config.StatusFailedMessage, reason)
}

func (argo *Argo) handleDeploymentSuccess(task m.Task) {
	log.Info().Str("id", task.Id).Msg("App is running on the excepted version.")
	failedDeployment.With(prometheus.Labels{"app": task.App}).Set(0)
	argo.state.SetTaskStatus(task.Id, "deployed", "")
}

func (argo *Argo) handleAppNotAvailable(task m.Task, err error) {
	log.Warn().Str("id", task.Id).Msgf("Deployment failed. Application not available\n%s", err.Error())
	failedDeployment.With(prometheus.Labels{"app": task.App}).Inc()
	reason := fmt.Sprintf("Application not available\n\n%s", err.Error())
	argo.state.SetTaskStatus(task.Id, config.StatusFailedMessage, reason)
}

func (argo *Argo) handleAppNotHealthy(task m.Task, err error) {
	log.Warn().Str("id", task.Id).Msgf("Deployment failed. Application not healthy\n%s", err.Error())
	failedDeployment.With(prometheus.Labels{"app": task.App}).Inc()
	reason := fmt.Sprintf("Application not healthy\n\n%s", err.Error())
	argo.state.SetTaskStatus(task.Id, config.StatusFailedMessage, reason)
}

func (argo *Argo) handleAppOutOfSync(task m.Task, err error) {
	log.Warn().Str("id", task.Id).Msgf("Deployment failed. Application out of sync\n%s", err.Error())
	failedDeployment.With(prometheus.Labels{"app": task.App}).Inc()
	reason := fmt.Sprintf("Application out of sync\n\n%s", err.Error())
	argo.state.SetTaskStatus(task.Id, config.StatusFailedMessage, reason)
}

func (argo *Argo) handleDeploymentUnexpectedStatus(task m.Task, err error) {
	log.Error().Str("id", task.Id).Msg("Deployment timed out with unexpected status. Aborting.")
	log.Error().Str("id", task.Id).Msgf("Deployment error\n%s", err.Error())
	failedDeployment.With(prometheus.Labels{"app": task.App}).Inc()
	reason := fmt.Sprintf("Deployment timeout\n\n%s", err.Error())
	argo.state.SetTaskStatus(task.Id, config.StatusFailedMessage, reason)
}

func (argo *Argo) GetAppList() []string {
	return argo.state.GetAppList()
}

func (argo *Argo) SimpleHealthCheck() bool {
	return argo.state.Check()
}
